File: wadup-web/docker/python/wadup-guest-python/wadup.py
"""WADUP Guest Library for Python modules.

This module provides the API for WADUP modules to define metadata tables
and insert extracted data.
"""
import json
import os
from typing import List, Tuple, Any, Dict

# Type mapping
TYPE_MAP = {
    "String": str,
    "Int64": int,
    "Float64": float,
    "Bool": bool,
    "Bytes": bytes,
}

# Global state for tables
_tables: Dict[str, "Table"] = {}


class Table:
    """Represents a metadata table."""

    def __init__(self, name: str, columns: List[Tuple[str, str]]):
        self.name = name
        self.columns = columns
        self.rows: List[List[Any]] = []

    def insert_row(self, values: List[Any]) -> None:
        """Insert a row into the table."""
        if len(values) != len(self.columns):
            raise ValueError(
                f"Expected {len(self.columns)} values, got {len(values)}"
            )
        self.rows.append(values)

    def to_dict(self) -> dict:
        """Convert table to dictionary for serialization."""
        return {
            "name": self.name,
            "columns": [{"name": c[0], "type": c[1]} for c in self.columns],
            "rows": self.rows,
        }


def define_table(name: str, columns: List[Tuple[str, str]]) -> Table:
    """Define a new metadata table.

    Args:
        name: The table name
        columns: List of (column_name, column_type) tuples
                Types: "String", "Int64", "Float64", "Bool", "Bytes"

    Returns:
        The created Table object
    """
    table = Table(name, columns)
    _tables[name] = table
    return table


def insert_row(table_name: str, values: List[Any]) -> None:
    """Insert a row into a table.

    Args:
        table_name: The name of the table
        values: List of values matching the table's column definitions
    """
    if table_name not in _tables:
        raise ValueError(f"Table '{table_name}' not defined")
    _tables[table_name].insert_row(values)


def flush() -> None:
    """Write all table data to the metadata output file."""
    output = {
        "tables": [table.to_dict() for table in _tables.values()]
    }

    # Write to metadata output file
    try:
        with open("/metadata.json", "w") as f:
            json.dump(output, f)
    except Exception as e:
        print(f"Warning: Failed to write metadata: {e}", file=__import__("sys").stderr)


def get_content_path() -> str:
    """Get the path to the input file content.

    Returns:
        Path to the input file (typically /data.bin)
    """
    return "/data.bin"


def get_filename() -> str:
    """Get the original filename of the input file.

    Returns:
        The filename from WADUP_FILENAME environment variable
    """
    return os.environ.get("WADUP_FILENAME", "unknown")

File: wadup-web/docker/go/wadup-guest-go/wadup.go
// Package wadup provides the WADUP guest library for Go modules.
package wadup

import (
	"encoding/json"
	"fmt"
	"os"
)

// DataType represents the type of a column value
type DataType int

const (
	String DataType = iota
	Int64
	Float64
	Bool
	Bytes
)

// Value represents a typed value for a table row
type Value struct {
	Type DataType
	data interface{}
}

// NewString creates a new string value
func NewString(s string) Value {
	return Value{Type: String, data: s}
}

// NewInt64 creates a new int64 value
func NewInt64(i int64) Value {
	return Value{Type: Int64, data: i}
}

// NewFloat64 creates a new float64 value
func NewFloat64(f float64) Value {
	return Value{Type: Float64, data: f}
}

// NewBool creates a new bool value
func NewBool(b bool) Value {
	return Value{Type: Bool, data: b}
}

// NewBytes creates a new bytes value
func NewBytes(b []byte) Value {
	return Value{Type: Bytes, data: b}
}

// Column represents a table column definition
type Column struct {
	Name string   `json:"name"`
	Type DataType `json:"type"`
}

// TableBuilder helps construct table definitions
type TableBuilder struct {
	name    string
	columns []Column
}

// Table represents a metadata table
type Table struct {
	Name    string   `json:"name"`
	Columns []Column `json:"columns"`
	Rows    [][]interface{} `json:"rows"`
}

// NewTableBuilder creates a new table builder
func NewTableBuilder(name string) *TableBuilder {
	return &TableBuilder{name: name, columns: []Column{}}
}

// Column adds a column to the table definition
func (tb *TableBuilder) Column(name string, dataType DataType) *TableBuilder {
	tb.columns = append(tb.columns, Column{Name: name, Type: dataType})
	return tb
}

// Build finalizes the table definition
func (tb *TableBuilder) Build() (*Table, error) {
	if tb.name == "" {
		return nil, fmt.Errorf("table name is required")
	}
	if len(tb.columns) == 0 {
		return nil, fmt.Errorf("at least one column is required")
	}

	table := &Table{
		Name:    tb.name,
		Columns: tb.columns,
		Rows:    [][]interface{}{},
	}

	// Register table globally
	tables = append(tables, table)

	return table, nil
}

// InsertRow inserts a row into the table
func (t *Table) InsertRow(values []Value) error {
	if len(values) != len(t.Columns) {
		return fmt.Errorf("expected %d values, got %d", len(t.Columns), len(values))
	}

	row := make([]interface{}, len(values))
	for i, v := range values {
		row[i] = v.data
	}
	t.Rows = append(t.Rows, row)
	return nil
}

// Global state
var tables []*Table

// Flush writes all metadata tables to the output file
func Flush() {
	output := map[string]interface{}{
		"tables": tables,
	}

	data, err := json.Marshal(output)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to marshal output: %v\n", err)
		return
	}

	// Write to the metadata output file
	err = os.WriteFile("/metadata.json", data, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write metadata: %v\n", err)
	}
}
